lib/spack/spack/package.py

```diff
@@ -2174,11 +2174,20 @@ def fetch_remote_versions(self):
             tty.die("Package.fetch_versions couldn't connect to:", e.url,
                     e.message)
 
+    @property
+    def rpath_deps(self):
+        """Return immediate or transitive RPATHs depending on the package."""
+        if self.transitive_rpaths:
+            return [d for d in self.spec.traverse(root=False,
+                                                  deptype=('link'))]
+        else:
+            return self.spec.dependencies(deptype='link')
+
     @property
     def rpath(self):
         """Get the rpath this package links with, as a list of paths."""
         rpaths = [self.prefix.lib, self.prefix.lib64]
-        deps = self.spec.dependencies(deptype='link')
```

Comment:
this is the only functional change in this PR.

Reply:
This does not appear to change the behavior of `get_rpaths`, but `Package.rpath_args` uses `Package.rpath`, so the functional change in this PR is to pull in dependency RPATHs transitively for packages using `rpath_args`; as you say, this is only when `Package.transitive_rpaths == True`, but that is the case by default (only a few disable `transitive_rpaths`).

To preserve the current logic (i.e. avoid any functional change and make this a "pure" refactor), I think `rpath` would have to become a function which accepts a `transitive` argument.

I'll note that only two packages (`gcc` and `tau`) use `rpath_args` (and nothing in `lib/spack/` uses it), so if we can determine there is no harm in both cases with this change in the logic, we wouldn't have to worry about it. I don't think there is an issue in `GCC`'s case other than that the argument would appear quite large in the build log and be harder to parse out by eye. This seems to be the case for `tau` as well.

On top of that, I generally I don't expect it to be a problem because the Spack compiler wrappers will insert RPATHs for transitive dependencies automatically.

Overall, would you agree?

Reply:
Yes, I agree.

Ps. dealii also sets transitive_rpath, but things still work, of course.

Reply:
@scheibelp do you insist on making this a pure refactor? I would keep the minor change in logic here as otherwise things become nasty and one needs to propagate that `transitive_rpaths` in all calls in both `Package::rpath()` as well as in `Package::rpath_args()`. I would appreciate if we could merge it as-is so that I go on with other PRs that are on hold for a half year due to this one...

Reply:
@scheibelp also note that the whole point of refactoring is to do the next-step fix in another PR https://github.com/spack/spack/pull/11342/commits/c0b9d081ebf34f2e1fb1bdfa4cb06dbb4e44cb92

Reply:
@scheibelp 

> I'll note that only two packages (gcc and tau) use rpath_args (and nothing in lib/spack/ uses it), so if we can determine there is no harm in both cases with this change in the logic, we wouldn't have to worry about it. 

on the rebased branch and clean Spack (no packages installed) I tried:
```
spack install gcc@9.2.0
spack install dealii@9.1.1
```

I can't build `tau` on macOS to try: `"platform=darwin" conflicts with "libunwind" [Non-GNU libunwind needs ELF libraries Darwin does not have]`. 

Bottom line: I would appreciate if this is reviewed/merged so that I clean up / finish up old PRs.

Reply:
> scheibelp on 19 Jun

@scheibelp ping.

Reply:
> do you insist on making this a pure refactor? I would keep the minor change in logic here as otherwise things become nasty and one needs to propagate that transitive_rpaths in all calls in both Package::rpath() as well as in Package::rpath_args()

What would need to be changed if `gcc`/`tau` end up requiring the old behavior? My original suggestion isn't required, but one of the following needs to happen:

* Build `gcc` and `tau` to confirm they aren't affected by this change
* Provide an additional alternative method that pulls direct dependencies rather than transitive, then use that method in `gcc`/`tau`
* Or: implement the suggestion in https://github.com/spack/spack/pull/10992/files#r295063046